### PR TITLE
Add new remote store options to Snapshot API

### LINKED ISSUE #4600: [DOC] Update Snapshot Restore API Documentation with Snapshot Interoperability w/ Remote Store Changes.
**What do you want to do?**
 
- [x] Request a change to existing documentation
- [ ] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.
 For our new feature Snapshot Interoperability w/ Remote Store, we need to update the [request fields](https://opensearch.org/docs/latest/api-reference/snapshots/restore-snapshot/#request-fields) for Restore Snapshot API with a new field `remote_store_index_shallow_copy`:

Parameter | Data type | Description
-- | -- | --
source_remote_store_repository | String | Remote Store Repository name of source indices which are being restored, if it was modified after snapshot creation. Optional.

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.
For testing, follow the steps mentioned in this doc to create shallow copy snapshot: https://github.com/opensearch-project/documentation-website/issues/4599.  
After taking snapshot, change repository name for remote store repo. Now, normal snapshot would fail for remote store indices, but if the updated repository name is passed as shown in below example, snapshot restore would work fine.

**Sample Request:**
```
curl -X POST "localhost:9200/_snapshot/shallowCopyRepository/shallow_snap/_restore?wait_for_completion=true" -H 'Content-Type: application/json' -d'
{
  "indices": "testidx-dr",
  "rename_pattern": "(.+)",
  "rename_replacement": "restored_shallow-$0",
  "source_remote_store_repository": "remoteStoreRepoUpdated"
}
'
```

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
Fixes #4600, #4599, and #4598

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _api-reference/snapshots/create-repository.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ Request field | Description
 `compress` | Whether to compress metadata files. This setting does not affect data files, which might already be compressed, depending on your index settings. Default is `false`. Optional.
 `max_restore_bytes_per_sec` | The maximum rate at which snapshots restore. Default is 40 MB per second (`40m`). Optional.
 `max_snapshot_bytes_per_sec` | The maximum rate at which snapshots take. Default is 40 MB per second (`40m`). Optional.
+`remote_store_index_shallow_copy` | Boolean | Whether or not the current snapshot is a full copy or shallow copy.
```
Comment (harishbhakuni) [inline _api-reference/snapshots/create-repository.md:46]: Customer needs to set this flag during create repository which when enabled, we will take snapshots of all the remote store enabled indices as shallow copy. maybe rephrase it to something like `Whether or not the snapshot of remote store enabled indices be captured as shallow copy.'

File: _api-reference/snapshots/create-repository.md [hunk 2]
```diff
@@ -74,6 +75,7 @@ Request field | Description
 `max_restore_bytes_per_sec` | The maximum rate at which snapshots restore. Default is 40 MB per second (`40m`). Optional.
 `max_snapshot_bytes_per_sec` | The maximum rate at which snapshots take. Default is 40 MB per second (`40m`). Optional.
 `readonly` | Whether the repository is read-only. Useful when migrating from one cluster (`"readonly": false` when registering) to another cluster (`"readonly": true` when registering). Optional.
+`remote_store_index_shallow_copy` | Boolean | Whether or not the current snapshot is a full copy or shallow copy.
```
Comment (harishbhakuni) [inline _api-reference/snapshots/create-repository.md:78]: here as well: `Whether or not the snapshot of remote store enabled indices be captured as shallow copy.' also, we can add that the default value of this will be False.

File: _api-reference/snapshots/create-repository.md (hunk 1 above)
Comment (Naarcha-AWS) [inline _api-reference/snapshots/create-repository.md:46]:
```suggestion
`remote_store_index_shallow_copy` | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy.
```

File: _api-reference/snapshots/create-repository.md (hunk 2 above)
Comment (Naarcha-AWS) [inline _api-reference/snapshots/create-repository.md:78]:
```suggestion
`remote_store_index_shallow_copy` | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`.
```

File: _api-reference/snapshots/create-snapshot.md [hunk 3]
```diff
@@ -141,4 +141,5 @@ The snapshot definition is returned.
 | duration_in_millis | long | Total time (in milliseconds) that the snapshot creation process lasted. |
 | failures | array | Failures, if any, that occured during snapshot creation. |
 | shards | object | Total number of shards created along with number of successful and failed shards. |
-| state | string | Snapshot status. Possible values: `IN_PROGRESS`, `SUCCESS`, `FAILED`, `PARTIAL`. |
\ No newline at end of file
+| state | string | Snapshot status. Possible values: `IN_PROGRESS`, `SUCCESS`, `FAILED`, `PARTIAL`. |
+| remote_store_index_shallow_copy | boolean | Whether or not the current snapshot is a full copy or shallow copy. |
```
Comment (Naarcha-AWS) [inline _api-reference/snapshots/create-snapshot.md:145]:
```suggestion
| remote_store_index_shallow_copy | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`. |
```

File: _api-reference/snapshots/restore-snapshot.md [hunk 4]
```diff
@@ -45,6 +45,7 @@ All request body parameters are optional.
...
 | rename_pattern | String | The pattern to apply to restored data streams and indices. Data streams and indices matching the rename pattern will be renamed according to `rename_replacement`. <br /><br />The rename pattern is applied as defined by the regular expression that supports referencing the original text. <br /> <br />The request fails if two or more data streams or indices are renamed into the same name. If you rename a restored data stream, its backing indices are also renamed. For example, if you rename the logs data stream to `recovered-logs`, the backing index `.ds-logs-1` is renamed to `.ds-recovered-logs-1`. <br /> <br />If you rename a restored stream, ensure an index template matches the new stream name. If there are no matching index template names, the stream cannot roll over and new backing indices are not created.|
 | rename_replacement | String | The rename replacement string. See `rename_pattern` for more information.|
+| source_remote_store_repository | String | The name of the remote store repository's source indexes that are being restored.
```
Comment (Naarcha-AWS) [inline _api-reference/snapshots/restore-snapshot.md:48]:
```suggestion
| source_remote_store_repository | String | The name of the remote store repository source index that is being restored.
```

**Comment by Naarcha-AWS** [comment]
@harishbhakuni: Adjusted.

File: _api-reference/snapshots/restore-snapshot.md [hunk 5]
```diff
@@ -45,6 +45,7 @@ All request body parameters are optional.
...
 | rename_pattern | String | The pattern to apply to restored data streams and indices. Data streams and indices matching the rename pattern will be renamed according to `rename_replacement`. <br /><br />The rename pattern is applied as defined by the regular expression that supports referencing the original text. <br /> <br />The request fails if two or more data streams or indices are renamed into the same name. If you rename a restored data stream, its backing indices are also renamed. For example, if you rename the logs data stream to `recovered-logs`, the backing index `.ds-logs-1` is renamed to `.ds-recovered-logs-1`. <br /> <br />If you rename a restored stream, ensure an index template matches the new stream name. If there are no matching index template names, the stream cannot roll over and new backing indices are not created.|
 | rename_replacement | String | The rename replacement string. See `rename_pattern` for more information.|
+| source_remote_store_repository | String | The name of the remote store repository source index that is being restored.
```
Comment (Naarcha-AWS) [inline _api-reference/snapshots/restore-snapshot.md:48]:
```suggestion
| source_remote_store_repository | String | The name of the remote store repository of source index that is being restored. If not provided, the repository that was registered when the snapshot was created will be used.
```

File: _api-reference/snapshots/create-repository.md [hunk 6]
```diff
@@ -43,6 +43,7 @@ Request field | Description
 `compress` | Whether to compress metadata files. This setting does not affect data files, which might already be compressed, depending on your index settings. Default is `false`. Optional.
 `max_restore_bytes_per_sec` | The maximum rate at which snapshots restore. Default is 40 MB per second (`40m`). Optional.
 `max_snapshot_bytes_per_sec` | The maximum rate at which snapshots take. Default is 40 MB per second (`40m`). Optional.
+`remote_store_index_shallow_copy` | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`.
```
Comment (cwillum) [inline _api-reference/snapshots/create-repository.md:46]: just suggesting to add a verb. Global if you decide it's necessary.
```suggestion
`remote_store_index_shallow_copy` | Boolean | Determines whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`.
```

File: _api-reference/snapshots/restore-snapshot.md [hunk 7]
```diff
@@ -45,6 +45,7 @@ All request body parameters are optional.
...
 | rename_pattern | String | The pattern to apply to restored data streams and indices. Data streams and indices matching the rename pattern will be renamed according to `rename_replacement`. <br /><br />The rename pattern is applied as defined by the regular expression that supports referencing the original text. <br /> <br />The request fails if two or more data streams or indices are renamed into the same name. If you rename a restored data stream, its backing indices are also renamed. For example, if you rename the logs data stream to `recovered-logs`, the backing index `.ds-logs-1` is renamed to `.ds-recovered-logs-1`. <br /> <br />If you rename a restored stream, ensure an index template matches the new stream name. If there are no matching index template names, the stream cannot roll over and new backing indices are not created.|
 | rename_replacement | String | The rename replacement string. See `rename_pattern` for more information.|
+| source_remote_store_repository | String | The name of the remote store repository of source index that is being restored. If not provided, the repository that was registered when the snapshot was created will be used.
```
Comment (cwillum) [inline _api-reference/snapshots/restore-snapshot.md:48]:
```suggestion
| source_remote_store_repository | String | The name of the remote store repository of the source index that is being restored. If not provided, the repository that was registered when the snapshot was created will be used.
```

**Review by cwillum — APPROVED** [review]
Looks good.

File: _api-reference/snapshots/create-repository.md (hunk 6 above)
Comment (natebower) [inline _api-reference/snapshots/create-repository.md:46]: "is" captured? (should agree with "the snapshot")

File: _api-reference/snapshots/create-repository.md [hunk 8]
```diff
@@ -74,6 +75,7 @@ Request field | Description
 `max_restore_bytes_per_sec` | The maximum rate at which snapshots restore. Default is 40 MB per second (`40m`). Optional.
 `max_snapshot_bytes_per_sec` | The maximum rate at which snapshots take. Default is 40 MB per second (`40m`). Optional.
 `readonly` | Whether the repository is read-only. Useful when migrating from one cluster (`"readonly": false` when registering) to another cluster (`"readonly": true` when registering). Optional.
+`remote_store_index_shallow_copy` | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`.
```
Comment (natebower) [inline _api-reference/snapshots/create-repository.md:78]: Same

File: _api-reference/snapshots/create-snapshot.md [hunk 9]
```diff
@@ -141,4 +141,5 @@ The snapshot definition is returned.
 | duration_in_millis | long | Total time (in milliseconds) that the snapshot creation process lasted. |
 | failures | array | Failures, if any, that occured during snapshot creation. |
 | shards | object | Total number of shards created along with number of successful and failed shards. |
-| state | string | Snapshot status. Possible values: `IN_PROGRESS`, `SUCCESS`, `FAILED`, `PARTIAL`. |
\ No newline at end of file
+| state | string | Snapshot status. Possible values: `IN_PROGRESS`, `SUCCESS`, `FAILED`, `PARTIAL`. |
+| remote_store_index_shallow_copy | Boolean | Whether the snapshot of the remote store indexes are captured as a shallow copy. Default is `false`. |
```
Comment (natebower) [inline _api-reference/snapshots/create-snapshot.md:145]: Same

File: _api-reference/snapshots/restore-snapshot.md [hunk 10]
```diff
@@ -45,6 +45,7 @@ All request body parameters are optional.
...
 | rename_pattern | String | The pattern to apply to restored data streams and indices. Data streams and indices matching the rename pattern will be renamed according to `rename_replacement`. <br /><br />The rename pattern is applied as defined by the regular expression that supports referencing the original text. <br /> <br />The request fails if two or more data streams or indices are renamed into the same name. If you rename a restored data stream, its backing indices are also renamed. For example, if you rename the logs data stream to `recovered-logs`, the backing index `.ds-logs-1` is renamed to `.ds-recovered-logs-1`. <br /> <br />If you rename a restored stream, ensure an index template matches the new stream name. If there are no matching index template names, the stream cannot roll over and new backing indices are not created.|
 | rename_replacement | String | The rename replacement string. See `rename_pattern` for more information.|
+| source_remote_store_repository | String | The name of the remote store repository of the source index that is being restored. If not provided, the repository that was registered when the snapshot was created will be used.
```
Comment (natebower) [inline _api-reference/snapshots/restore-snapshot.md:48]:
```suggestion
| source_remote_store_repository | String | The name of the remote store repository of the source index being restored. If not provided, the repository that was registered when the snapshot was created will be used.
```

Comment (natebower) [inline _api-reference/snapshots/restore-snapshot.md:48]: Would the second sentence be better (more active) as "If not provided, [noun] will use the repository that was registered when the snapshot was created"?

**Review by natebower — COMMENTED** [review]
@Naarcha-AWS Just a few changes. Be sure to apply changes consistently across each instance. Thanks!

File: _api-reference/snapshots/create-repository.md (hunk 8 above)
Comment (Naarcha-AWS) [inline _api-reference/snapshots/create-repository.md:78]:
```suggestion
`remote_store_index_shallow_copy` | Boolean | Whether the snapshot of the remote store indexes is captured as a shallow copy. Default is `false`.
```

File: _api-reference/snapshots/create-snapshot.md (hunk 9 above)
Comment (Naarcha-AWS) [inline _api-reference/snapshots/create-snapshot.md:145]:
```suggestion
| remote_store_index_shallow_copy | Boolean | Whether the snapshot of the remote store indexes is captured as a shallow copy. Default is `false`. |
```

File: _api-reference/snapshots/restore-snapshot.md (hunk 10 above)
Comment (Naarcha-AWS) [reply _api-reference/snapshots/restore-snapshot.md:48]: Snapshot restore API
